libqtile/layout/plasma.py

```diff
@@ -0,0 +1,1008 @@
+        flip = False
+        if len(self.siblings) == 1:
+            # If there is only one node left in the container, we need to save
+            # its size too because the size will be lost.
+            sizes += (self.siblings[0]._size,)  # pylint: disable=W0212
+            if not self.parent.is_root:
+                flip = True
+                parent = self.siblings[0]
+        self.root.restorables[self.payload] = (parent, self.index, sizes, self.fixed, flip)
+
+    def move(self, direction):
+        """Move this node in `direction`. Return whether node was moved."""
+        if self.is_root:
+            return False
+        if direction.orient is self.parent.orient:
+            old_idx = self.index
+            new_idx = old_idx + direction.offset
+            if 0 <= new_idx < len(self.parent):
+                p = self.parent
+                p[old_idx], p[new_idx] = p[new_idx], p[old_idx]
+                return True
+            new_sibling = self.parent.parent
+        else:
+            new_sibling = self.parent
+        try:
+            new_parent = new_sibling.parent
+            idx = new_sibling.index
+        except AttributeError:
+            return False
+        self.reset_size()
+        self.parent.remove_child(self)
+        new_parent.add_child(self, idx + (1 if direction.offset == 1 else 0))
+        return True
+
+    def move_up(self):
+        return self.move(Direction.UP)
+
+    def move_down(self):
+        return self.move(Direction.DOWN)
+
+    def move_right(self):
+        return self.move(Direction.RIGHT)
+
+    def move_left(self):
+        return self.move(Direction.LEFT)
+
+    def _move_and_integrate(self, direction):
+        old_parent = self.parent
+        self.move(direction)
+        if self.parent is not old_parent:
+            self.integrate(direction)
+
+    def integrate(self, direction):
+        if direction.orient != self.parent.orient:
+            self._move_and_integrate(direction)
+            return
+        target_idx = self.index + direction.offset
+        if target_idx < 0 or target_idx >= len(self.parent):
+            self._move_and_integrate(direction)
+            return
+        self.reset_size()
+        target = self.parent[target_idx]
+        self.parent.remove_child(self)
+        if target.is_leaf:
+            target.flip_with(self)
+        else:
+            target.add_child(self)
+
+    def integrate_up(self):
+        self.integrate(Direction.UP)
+
+    def integrate_down(self):
+        self.integrate(Direction.DOWN)
+
+    def integrate_left(self):
+        self.integrate(Direction.LEFT)
+
+    def integrate_right(self):
+        self.integrate(Direction.RIGHT)
+
+    def find_payload(self, payload):
+        if self.payload is payload:
+            return self
+        for child in self:
+            needle = child.find_payload(payload)
+            if needle is not None:
+                return needle
+        return None
+
+
+class Plasma(Layout):
```

Comment:
Maybe add the fact that this was forked and upstreames from the original qtile plasma layout with a link to it. That way some better credit is given apart from the copyright header on the top.

Reply:
Agreed. I want to add much more on documentation, docstrings, comments etc before merging

Reply:
Thanks! I was shocked at the line count at first but 1.6k in tests is really nice! I would say 1k for the layout is manageable. So I think this is indeed a good candidate for up streaming, especially because a bunch of users started with i3

Reply:
Yeah. The amount of tests here is really impressive. We've got to 100% coverage of the layout.